docs/proposals/blocks-storage-sharding.md

```diff
@@ -0,0 +1,84 @@
+---
+title: "Blocks storage sharding"
+linkTitle: "Blocks storage sharding"
+weight: 1
+slug: blocks-storage-sharding
+---
+
+- Author: [Marco Pracucci](https://github.com/pracucci)
+- Date: March 2020
+- Status: proposal
+
+## Problem
+
+In Cortex, when using the experimental blocks storage, each querier internally runs the Thanos [`BucketStore`](https://github.com/thanos-io/thanos/blob/master/pkg/store/bucket.go). This means that each querier has a full view over all blocks in the long-term storage and all blocks index headers are loaded in each querier memory. The querier memory usage linearly increase with number and size of all blocks in the storage, imposing a scalability limit to the blocks storage.
+
+In this proposal we want to solve this. In particular, we want to:
+
+1. Shard blocks (index headers) across a pool of nodes
+2. Do not compromise HA on the read path (if a node fails, queries should continue to work)
+3. Do not compromise correctness (either the query result is correct or it fails)
+
+## Proposed solution
+
+The idea is to introduce a new Cortex service - `store-gateway` - internally running the Thanos [`BucketStore`](https://github.com/thanos-io/thanos/blob/master/pkg/store/bucket.go). At query time, a querier will run a query fetching the matching series both from ingesters and the subset of gateways holding the related blocks (based on the query time range). Blocks are replicated across the gateways in order to guarantee query results consistency and HA even in the case of a gateway instance failure.
```

Comment:
Is there a reason why it needs to be a new service? Could we not simply have each querier maintain it's own shard of blocks, and then ask other queriers when necessary?

Reply:
Perhaps. Having it as a separate component allows independent scaling of queriers and store gateways. If we later realize that they can run together, we can simply use multiple targets when starting them (support for that is coming soon).

Reply:
+1 for a separate component. I can see this running in-process with the querier with multi-targeting on smaller setups but completely separate on larger configurations.